pkg/kubectl/cmd/util/clientcache.go

```diff
@@ -29,18 +31,39 @@ import (
 	"k8s.io/kubernetes/pkg/runtime/schema"
 )
 
-func NewClientCache(loader clientcmd.ClientConfig) *ClientCache {
-	return &ClientCache{
+const (
+	FlagMatchBinaryVersion = "match-server-version"
+)
+
+// ClientCache caches previously loaded clients for reuse, and ensures MatchServerVersion
+// is invoked only once
+type ClientCache interface {
+	// ClientConfigForVersion returns the correct config for a server
+	ClientConfigForVersion(requiredVersion *schema.GroupVersion) (*restclient.Config, error)
+	// ClientSetForVersion initializes or reuses a clientset for the specified version, or returns an
+	// error if that is not possible
+	ClientSetForVersion(requiredVersion *schema.GroupVersion) (*internalclientset.Clientset, error)
```

Comment:
Isn't the expensive part the negotiation and transports get re-used at a lower level based on config compatibility?  If so, we shouldn't leak this.